File: examples/csharp/Logos.Payment.Service.CliHost/CliParser.cs
namespace Logos.Payment.Service.CliHost;

public class CliParser
{
    private readonly string _command;
    private readonly List<string> _positionalArgs = new();
    private readonly Dictionary<string, string> _options = new();
    private readonly HashSet<string> _flags = new();

    public CliParser(string[] args)
    {
        if (args.Length < 1)
        {
            _command = string.Empty;
            return;
        }

        _command = args[0];

        for (int i = 1; i < args.Length; i++)
        {
            string arg = args[i];

            if (arg.StartsWith("--"))
            {
                string optionName = arg.Substring(2);

                // Check if next argument is a value (doesn't start with --)
                if (i + 1 < args.Length && !args[i + 1].StartsWith("--"))
                {
                    _options[optionName] = args[i + 1];
                    i++; // Skip the value
                }
                else
                {
                    // It's a flag
                    _flags.Add(optionName);
                }
            }
            else
            {
                _positionalArgs.Add(arg);
            }
        }
    }

    public string GetCommand() => _command;

    public bool HasFlag(string flag) => _flags.Contains(flag);

    public string GetOption(string option) => _options.TryGetValue(option, out var value) ? value : string.Empty;

    public string GetPositional(int index) => index < _positionalArgs.Count ? _positionalArgs[index] : string.Empty;

    public bool IsValid() => !string.IsNullOrEmpty(_command);
}

File: examples/cpp/mypaymentservice_cli_host/cli_parser.cpp
#include "mypaymentservice_cli_host/cli_parser.h"
#include <algorithm>

namespace mypaymentservice::cli_host {

CliParser::CliParser(int argc, char* argv[]) {
    if (argc < 2) {
        return;
    }

    command_ = argv[1];

    for (int i = 2; i < argc; ++i) {
        std::string arg = argv[i];

        if (arg.substr(0, 2) == "--") {
            // Option with value: --amount 100
            if (i + 1 < argc && argv[i + 1][0] != '-') {
                options_[arg.substr(2)] = argv[i + 1];
                ++i;
            } else {
                // Flag: --help
                flags_.push_back(arg.substr(2));
            }
        } else {
            // Positional argument
            positional_args_.push_back(arg);
        }
    }
}

std::string CliParser::GetCommand() const {
    return command_;
}

bool CliParser::HasFlag(const std::string& flag) const {
    return std::find(flags_.begin(), flags_.end(), flag) != flags_.end();
}

std::string CliParser::GetOption(const std::string& option) const {
    auto it = options_.find(option);
    return it != options_.end() ? it->second : "";
}

std::string CliParser::GetPositional(size_t index) const {
    return index < positional_args_.size() ? positional_args_[index] : "";
}

bool CliParser::IsValid() const {
    return !command_.empty();
}

} // namespace mypaymentservice::cli_host
